### PR TITLE
Bump dependencies, add missing typing libraries

### PR DESCRIPTION
**Pull request by davwheat**
**Changes proposed in this pull request:**
- Bumps dependencies to their latest versions in `package.json`.
  - ~~Mostly of interest is `spin.js` which has had a [few breaking changes](https://github.com/fgnass/spin.js/blob/master/CHANGELOG.md#400---2018-05-06), but they shouldn't affect us. They actually improve performance :)~~
    - It might be worth dropping this in favour of a native CSS-only solution as part of flarum/issue-archive#179, saving on [about 1.5 kB of JS](https://bundlephobia.com/result?p=spin.js@4.1.0)
  - I've also bumped Prettier and made sure it's now auto-updating to `^2.2.1` as before it was fixed to `2.0.2` (was this intentional?)
  - `expose-loader` has also been updated and had breaking changes, but these have been fixed as part of the upgrade process
- Adds missing typing libraries
  - `@types/jquery`
    - Remove export from our typing lib as `@types/jquery` should be used by extensions (and because re-exporting it causes issues)
  - `@types/lodash-es`
  - `@types/punycode`
  - `@types/textarea-caret`
- Moves `devdeps` that were incorrectly listed under `deps`
  - `webpack` (and related)
  - `@types/mithril`
  - `@babel/preset-typescript`
- Use lockfile version 2 syntax (npm 7+)

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.

### CODE REVIEW

**Comment by davwheat** [comment]
Woop woop -147 bytes of JS 😂

File: js/src/common/index.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
-import 'expose-loader?$!expose-loader?jQuery!jquery';
-import 'expose-loader?m!mithril';
-import 'expose-loader?dayjs!dayjs';
+// Expose jQuery, mithril and dayjs to the window browser object
+import 'expose-loader?exposes[]=$&exposes[]=jQuery!jquery';
+import 'expose-loader?exposes=m!mithril';
```
Comment (askvortsov1) [inline js/src/common/index.js:3]: This is a nicer API, I like this change

File: less/common/LoadingIndicator.less [hunk 2]
```diff
@@ -12,3 +12,50 @@
 .LoadingIndicator--block {
   height: 100px;
 }
+
```
Comment (askvortsov1) [inline less/common/LoadingIndicator.less:15]: Is this the CSS-only solution? If so, it should be a separate PR.

**Review by askvortsov1 — CHANGES_REQUESTED** [review]

File: js/package.json [hunk 3]
```diff
@@ -2,29 +2,33 @@
   "private": true,
   "name": "@flarum/core",
   "dependencies": {
-    "@babel/preset-typescript": "^7.10.1",
-    "@types/mithril": "^2.0.3",
     "bootstrap": "^3.4.1",
-    "classnames": "^2.2.5",
+    "classnames": "^2.3.1",
```
[the remark this answers is not in the document]
Comment (davwheat) [reply js/package.json:6]: My plan was that, while it's in, we might as well update it. I was actually planning to submit the `clsx` PR tonight :P

File: less/common/LoadingIndicator.less (hunk 2 above)
Comment (davwheat) [reply less/common/LoadingIndicator.less:15]: No, these are the styles from `spin.js`. Before, they were dynamically injected when the JS was loaded (or called?), but now they're split.

They either need to be included with `import "node_modules/spin.js/style.css"`, or this methods. I don't think our webpack setup allows for important stylesheets, so I just added it here, as it was the most logical place.

File: js/src/common/index.js (hunk 1 above)
Comment (davwheat) [reply js/src/common/index.js:3]: Me too! v2 has an even nicer API too!

```js
import 'expose-loader?exposes=$,jQuery!jquery';
```

Comment (askvortsov1) [reply js/src/common/index.js:3]: > Me too! v2 has an even nicer API too, but that's Webpack 5 only :(

We'll need to update to Webpack 5 for the automated export stuff. Supporting it isn't difficult at all either, see https://github.com/flarum/core/compare/as/support-webpack-5

File: less/common/LoadingIndicator.less (hunk 2 above)
Comment (askvortsov1) [reply less/common/LoadingIndicator.less:15]: Can we skip updating spin.js then? I'd rather revisit it as a whole in #2345

**Review by askvortsov1 — APPROVED** [review]

**Comment by davwheat** [comment]
Fixed the conflicts from `clsx` PR

**Review by tankerkiller125 — APPROVED** [review]

**Comment by davwheat** [comment]
I just suddenly noticed that I managed to accidentally keep spin.js updated! Must have forgotten to push the changes. I'll quickly push that and merge.
